TLS Handshake Instrumentation

carterkozak opened the pull request:
Adds a meter for metric name `tls.handshake` with tags `name`, `cipher`, and `protocol`.

Open questions:
- [x] Is Tritium the right place for this instrumentation?
- [x] Are there any additional metrics we want to capture?

n.b. Conscrypt uses `instanceof` checks against socket factories and engines, we may need to provide and unwrap function for anything specific to conscrypt. This should only be an issue for components using conscrypt for HTTP/2 on java 8.

carterkozak committed:
[WIP] Quick and dirty handshake instrumentation

TODO: investigate API changes between java 8 and 12, we don't want
to break ALPN on new JVMs

carterkozak committed:
SSLEngine java 9+ methods

carterkozak commented on tritium-metrics/src/main/java/com/palantir/tritium/metrics/tls/InstrumentedSslEngine.java:
```diff
@@ -0,0 +1,314 @@
+            engineMethodNullable("getHandshakeApplicationProtocolSelector");
+
+    // Override(java9+)
+    public String getApplicationProtocol() {
```
We're not the only ones who use reflection check for java 9+ methods. We should provide SslEngine8 and SslEngine9 implementations, only the latter should implement these.

carterkozak commented on tritium-metrics/src/main/java/com/palantir/tritium/metrics/tls/InstrumentedSslEngine.java:
```diff
@@ -0,0 +1,314 @@
+            engineMethodNullable("setHandshakeApplicationProtocolSelector", BiFunction.class);
+    @Nullable
+    private static final Method getHandshakeApplicationProtocolSelector =
+            engineMethodNullable("getHandshakeApplicationProtocolSelector");
```
Perhaps scrape these from the delegate `engine.getClass()` in case that implementation provides a newer implementation than the jvm

carterkozak commented on tritium-metrics/src/main/java/com/palantir/tritium/metrics/tls/InstrumentedSslEngine.java:
```diff
@@ -0,0 +1,314 @@
+    @Override
+    public void beginHandshake() throws SSLException {
+        engine.beginHandshake();
+        handshaking.set(true);
```
Needs a comment describing that this needs to be reset for renegotiation.

carterkozak commented on tritium-metrics/src/main/java/com/palantir/tritium/metrics/tls/TlsMetrics.java:
```diff
@@ -0,0 +1,52 @@
+
+/** TODO(ckozak): docs. */
+@Beta
+public final class TlsMetrics {
```
Delete and move instrument methods into the MetricRegistries utility class.

carterkozak committed:
Only apply java9+ SSLEngine methods when present on the delegate.

carterkozak committed:
document renegotiation

carterkozak committed:
Suppress unused

carterkozak committed:
Javadoc, move into main package

carterkozak committed:
Test instrumentation

Fix SSLEngine java9 ALPN methods accessibility

carterkozak committed:
Remove public instrumentation for SSLEngine and ServerSSLSocketFactory

Both of these can be created using SSLContext, which we already
expose instrumentation for.

Updated javadoc, added not-null validation on args.

carterkozak committed:
Remove the `type` tag, `name` is sufficient.

carterkozak commented on tritium-metrics/build.gradle:
```diff
@@ -26,6 +26,8 @@ dependencies {
     testImplementation 'org.assertj:assertj-core'
     testImplementation 'org.mockito:mockito-core'
+    testImplementation 'io.undertow:undertow-core'
+    testImplementation 'com.squareup.okhttp3:okhttp'
```
Are we okay with bringing these in for testing?

schlosna replied:
I'm fine with these as test dependencies

schlosna commented on tritium-metrics/src/main/java/com/palantir/tritium/metrics/InstrumentedSslSocketFactory.java:
```diff
@@ -0,0 +1,108 @@
+                .putSafeTags("cipher", event.getCipherSuite())
+                .putSafeTags("protocol", event.getSession().getProtocol())
+                .build())
+                .mark();
```
do you want to include a tag for `type`?

carterkozak replied:
Originally I included a type tag, but decided to remove it because I think the name tag will give us the information we want.

Removed in https://github.com/palantir/tritium/commit/a1b544721fa350072404c849149f036bf45d441e

schlosna replied:
the `String type` method parameter is unused now, should we remove it?

carterkozak replied:
doh, fixed. Thanks for catching.

schlosna commented on tritium-metrics/src/main/java/com/palantir/tritium/metrics/InstrumentedSslContext.java:
```diff
@@ -0,0 +1,118 @@
+
+        @Override
+        protected SSLEngine engineCreateSSLEngine(String host, int port) {
+            return InstrumentedSslEngine.instrument(context.createSSLEngine(host, port), metrics, name);
```
we can probably infer this from the handshakes, but it might be interesting to have a counter of live `SSLEngine`s that have not been closed inbound/outbound.

carterkozak replied:
That would be interesting. This PR has grown a bit larger than I intended, I can investigate adding a counter in a follow-up.

carterkozak committed:
Merge remote-tracking branch 'origin/develop' into ckozak/tls_instrumentation

carterkozak committed:
Test server http/2 instrumentation

Assert meters have count 1.

carterkozak committed:
Reflective operations use doPrivileged if a SecurityManager is present

sfackler commented on tritium-metrics/src/main/java/com/palantir/tritium/metrics/InstrumentedSslSocketFactory.java:
```diff
@@ -0,0 +1,108 @@
+                .safeName("tls.handshake")
+                .putSafeTags("name", name)
+                .putSafeTags("cipher", event.getCipherSuite())
+                .putSafeTags("protocol", event.getSession().getProtocol())
```
There was some discussion about tracking session resumption as well - is it possible to grab that info?

carterkozak replied:
Not easily, as far as I can tell. Since it's non-trivial I'd prefer to hold off for another PR to implement that feature.

sfackler replied:
👍

carterkozak committed:
remove unused type tag parameter

schlosna commented:
In terms of "is tritium the right place for this", it feels right but curious for folks thoughts about the potential for dependency bloat and how folks would want to do injection for the metrics registry.

Taking one [example `conjure-java-runtime:keystores` where I'm assuming we'd want to instrument `SslSocketFactories`](https://github.com/palantir/conjure-java-runtime/blob/develop/keystores/src/main/java/com/palantir/conjure/java/config/ssl/SslSocketFactories.java#L70), `conjure-java-runtime:keystores` runtime dependencies are relatively smaller than `tritium-metrics`.

There's also [atlasdb's Cassandra SSL socket factory](https://github.com/palantir/atlasdb/blob/a7d4fe25281e7814cb124c92c4476dfad6ca8277/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientFactory.java#L67) and [atlasdb's Feign SSL socket factory](https://github.com/palantir/atlasdb/blob/45262270daffc86e500bb4d3499cdd63dff450f6/atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FeignOkHttpClients.java#L133) though that should come from conjure.

```
# conjure-java-runtime:keystores runtime dependencies

runtimeClasspath - Runtime classpath of source set 'main'.
+--- com.fasterxml.jackson.core:jackson-databind:{prefer 2.9.8} -> 2.9.8
|    +--- com.fasterxml.jackson.core:jackson-annotations:2.9.0 -> 2.9.8
|    \--- com.fasterxml.jackson.core:jackson-core:2.9.8
+--- com.palantir.conjure.java.api:ssl-config:{prefer 2.3.0} -> 2.3.0
|    \--- com.fasterxml.jackson.core:jackson-databind:2.9.7 -> 2.9.8 (*)
+--- com.google.guava:guava -> 27.0.1-jre
|    +--- com.google.guava:failureaccess:1.0.1
|    +--- com.google.guava:listenablefuture:9999.0-empty-to-avoid-conflict-with-guava
|    +--- com.google.code.findbugs:jsr305:3.0.2
|    +--- org.checkerframework:checker-qual:2.5.2 -> 2.6.0
|    +--- com.google.errorprone:error_prone_annotations:2.2.0 -> 2.3.3
|    +--- com.google.j2objc:j2objc-annotations:1.1
|    \--- org.codehaus.mojo:animal-sniffer-annotations:1.17
+--- com.fasterxml.jackson.core:jackson-databind:{strictly 2.9.8} -> 2.9.8 (c)
+--- com.google.guava:guava:{strictly 27.0.1-jre} -> 27.0.1-jre (c)
+--- com.palantir.conjure.java.api:ssl-config:{strictly 2.3.0} -> 2.3.0 (c)
+--- com.google.guava:guava:23.6.1-jre -> 27.0.1-jre (c)
+--- com.fasterxml.jackson.core:jackson-annotations:{strictly 2.9.8} -> 2.9.8 (c)
+--- com.fasterxml.jackson.core:jackson-core:{strictly 2.9.8} -> 2.9.8 (c)
+--- com.google.code.findbugs:jsr305:{strictly 3.0.2} -> 3.0.2 (c)
+--- com.google.errorprone:error_prone_annotations:{strictly 2.3.3} -> 2.3.3 (c)
+--- com.google.guava:failureaccess:{strictly 1.0.1} -> 1.0.1 (c)
+--- com.google.guava:listenablefuture:{strictly 9999.0-empty-to-avoid-conflict-with-guava} -> 9999.0-empty-to-avoid-conflict-with-guava (c)
+--- com.google.j2objc:j2objc-annotations:{strictly 1.1} -> 1.1 (c)
+--- org.checkerframework:checker-qual:{strictly 2.6.0} -> 2.6.0 (c)
+--- org.codehaus.mojo:animal-sniffer-annotations:{strictly 1.17} -> 1.17 (c)
\--- com.google.code.findbugs:jsr305:3.0.2 (c)
```

```
# tritium-metrics runtime dependencies:

runtimeClasspath - Runtime classpath of source set 'main'.
+--- project :tritium-api
|    +--- com.google.code.findbugs:jsr305 -> 3.0.2
|    +--- io.dropwizard.metrics:metrics-core:3.2.5 (c)
|    +--- com.google.guava:guava:23.6.1-jre -> 27.0.1-jre (c)
|    +--- com.google.code.findbugs:jsr305:3.0.2 (c)
|    +--- org.checkerframework:checker-qual:2.5.0 -> 2.6.0 (c)
|    \--- com.google.errorprone:error_prone_annotations:2.3.3 (c)
+--- project :tritium-core
|    +--- project :tritium-api (*)
|    +--- com.google.guava:guava -> 27.0.1-jre
|    |    +--- com.google.guava:failureaccess:1.0.1
|    |    +--- com.google.guava:listenablefuture:9999.0-empty-to-avoid-conflict-with-guava
|    |    +--- com.google.code.findbugs:jsr305:3.0.2
|    |    +--- org.checkerframework:checker-qual:2.5.2 -> 2.6.0
|    |    +--- com.google.errorprone:error_prone_annotations:2.2.0 -> 2.3.3
|    |    +--- com.google.j2objc:j2objc-annotations:1.1
|    |    \--- org.codehaus.mojo:animal-sniffer-annotations:1.17
|    +--- com.palantir.safe-logging:preconditions:{prefer 1.8.2} -> 1.8.2
|    |    +--- com.palantir.safe-logging:safe-logging:1.8.2
|    |    \--- com.google.errorprone:error_prone_annotations:2.3.2 -> 2.3.3
|    +--- com.palantir.safe-logging:safe-logging:{prefer 1.8.2} -> 1.8.2
|    +--- org.slf4j:slf4j-api:{prefer 1.7.12} -> 1.7.22
|    +--- io.dropwizard.metrics:metrics-core:3.2.5 (c)
|    +--- com.google.guava:guava:23.6.1-jre -> 27.0.1-jre (c)
|    +--- com.google.code.findbugs:jsr305:3.0.2 (c)
|    +--- org.checkerframework:checker-qual:2.5.0 -> 2.6.0 (c)
|    \--- com.google.errorprone:error_prone_annotations:2.3.3 (c)
+--- project :tritium-registry
|    +--- com.google.guava:guava -> 27.0.1-jre (*)
|    +--- com.palantir.safe-logging:preconditions:{prefer 1.8.2} -> 1.8.2 (*)
|    +--- com.palantir.safe-logging:safe-logging:{prefer 1.8.2} -> 1.8.2
|    +--- io.dropwizard.metrics:metrics-core -> 3.2.5
|    |    \--- org.slf4j:slf4j-api:1.7.22
|    +--- io.dropwizard.metrics:metrics-core:3.2.5 (c)
|    +--- com.google.guava:guava:23.6.1-jre -> 27.0.1-jre (c)
|    +--- com.google.code.findbugs:jsr305:3.0.2 (c)
|    +--- org.checkerframework:checker-qual:2.5.0 -> 2.6.0 (c)
|    \--- com.google.errorprone:error_prone_annotations:2.3.3 (c)
+--- io.dropwizard.metrics:metrics-core -> 3.2.5 (*)
+--- com.google.guava:guava -> 27.0.1-jre (*)
+--- com.palantir.safe-logging:preconditions:{prefer 1.8.2} -> 1.8.2 (*)
+--- com.palantir.safe-logging:safe-logging:{prefer 1.8.2} -> 1.8.2
+--- org.mpierce.metrics.reservoir:hdrhistogram-metrics-reservoir:{prefer 1.1.2} -> 1.1.2
|    \--- com.google.code.findbugs:jsr305:3.0.1 -> 3.0.2
+--- org.hdrhistogram:HdrHistogram:{prefer 2.1.11} -> 2.1.11
+--- org.slf4j:slf4j-api:{prefer 1.7.12} -> 1.7.22
+--- com.google.guava:guava:{strictly 27.0.1-jre} -> 27.0.1-jre (c)
+--- com.palantir.safe-logging:preconditions:{strictly 1.8.2} -> 1.8.2 (c)
+--- com.palantir.safe-logging:safe-logging:{strictly 1.8.2} -> 1.8.2 (c)
+--- io.dropwizard.metrics:metrics-core:{strictly 3.2.5} -> 3.2.5 (c)
+--- org.hdrhistogram:HdrHistogram:{strictly 2.1.11} -> 2.1.11 (c)
+--- org.mpierce.metrics.reservoir:hdrhistogram-metrics-reservoir:{strictly 1.1.2} -> 1.1.2 (c)
+--- org.slf4j:slf4j-api:{strictly 1.7.22} -> 1.7.22 (c)
+--- io.dropwizard.metrics:metrics-core:3.2.5 (c)
+--- com.google.guava:guava:23.6.1-jre -> 27.0.1-jre (c)
+--- com.google.code.findbugs:jsr305:{strictly 3.0.2} -> 3.0.2 (c)
+--- com.google.code.findbugs:jsr305:3.0.2 (c)
+--- com.google.errorprone:error_prone_annotations:{strictly 2.3.3} -> 2.3.3 (c)
+--- com.google.guava:failureaccess:{strictly 1.0.1} -> 1.0.1 (c)
+--- com.google.guava:listenablefuture:{strictly 9999.0-empty-to-avoid-conflict-with-guava} -> 9999.0-empty-to-avoid-conflict-with-guava (c)
+--- com.google.j2objc:j2objc-annotations:{strictly 1.1} -> 1.1 (c)
+--- org.checkerframework:checker-qual:{strictly 2.6.0} -> 2.6.0 (c)
+--- org.codehaus.mojo:animal-sniffer-annotations:{strictly 1.17} -> 1.17 (c)
+--- org.checkerframework:checker-qual:2.5.0 -> 2.6.0 (c)
\--- com.google.errorprone:error_prone_annotations:2.3.3 (c)
```

carterkozak commented:
Regarding dependency bloat, we'll need a dependency on `tritium-registry` if we want to produce tagged metrics. `tritium-metrics` and `tritium-core` don't add a great deal of bloat beyond the registry dependencies.

We could decouple this instrumentation into a framework to add java TLS callbacks, but at the end of day we would have tritium on the classpath for instrumentation (or we wouldn't, in which case we would not have metrics)

schlosna committed:
Merge branch 'develop' into ckozak/tls_instrumentation

carterkozak committed:
Allow access to wrapped SSLEngine delegate instances (#248)

* Allow access too wrapped SSLEngine delegate instances

This is necessary for servers which provide Conscrypt-based ALPN
using Conscryp on java 8:
https://github.com/google/conscrypt/blob/d7d3ec873a50c733c9c4397d6718c04045b377b3/common/src/main/java/org/conscrypt/Conscrypt.java#L638-L649

* null checks
